File: Frontend/src/types/contracts/practice.ts
import { z } from "zod";

import { difficultySchema, interviewTypeSchema } from "@/types/contracts/common";
import { historyMetricToneSchema } from "@/types/contracts/history";

export const practiceConfigSchema = z.object({
  role: z.string(),
  company: z.string(),
  type: interviewTypeSchema,
  difficulty: difficultySchema,
  duration: z.number(),
  focusAreas: z.array(z.string()),
  interviewerStyle: z.string(),
});

export const questionRefSchema = z.object({
  id: z.string(),
  text: z.string(),
  category: z.string(),
  topic: z.string(),
  difficulty: difficultySchema,
  followUp: z.boolean().optional(),
});

export const sessionAnswerSchema = z.object({
  questionId: z.string(),
  question: z.string(),
  transcript: z.string(),
  durationSeconds: z.number(),
  score: z.number(),
});

export const practiceSessionSchema = z.object({
  id: z.string(),
  status: z.enum(["ready", "active", "processing", "completed"]),
  config: practiceConfigSchema,
  questions: z.array(questionRefSchema),
  currentQuestionIndex: z.number(),
  answers: z.array(sessionAnswerSchema),
  startedAt: z.string().optional(),
});

export const answerReviewSchema = z.object({
  question: z.string(),
  answer: z.string(),
  score: z.number(),
  strengths: z.array(z.string()),
  missing: z.array(z.string()),
  betterStructure: z.array(z.string()),
});

const speechMetricsSchema = z.object({
  averageWpm: z.number(),
  fillerCount: z.number(),
  fillers: z.record(z.string(), z.number()),
  longPauses: z.number(),
  longestPause: z.number(),
  averageAnswerSeconds: z.number(),
});

export const interviewReportSchema = z.object({
  id: z.string(),
  sessionId: z.string(),
  createdAt: z.string(),
  overall: z.number(),
  technical: z.number(),
  communication: z.number(),
  structure: z.number(),
  clarity: z.number(),
  relevance: z.number(),
  depth: z.number(),
  summary: z.string(),
  speech: speechMetricsSchema,
  weakTopics: z.array(z.string()),
  strengths: z.array(z.string()),
  recommendedActions: z.array(z.string()),
  answers: z.array(answerReviewSchema),
});

export const completionSignatureAxisSchema = z.object({
  key: z.string(),
  label: z.string(),
  value: z.number(),
  benchmark: z.number(),
});

export const completionMetricSchema = z.object({
  key: z.string(),
  label: z.string(),
  value: z.number(),
  band: z.string(),
  tone: historyMetricToneSchema,
  delta: z.string().nullable(),
});

export const growthProtocolSchema = z.object({
  id: z.string(),
  priority: z.enum(["high", "medium", "low"]),
  title: z.string(),
  detail: z.string(),
  focusArea: z.string(),
});

export const completionQuestionSchema = z.object({
  id: z.string(),
  position: z.number(),
  question: z.string(),
  topic: z.string(),
  category: z.string(),
  difficulty: difficultySchema,
  score: z.number(),
  durationSeconds: z.number(),
  verdict: z.enum(["strong", "solid", "needs_work"]),
  answer: z.string(),
  strengths: z.array(z.string()),
  missing: z.array(z.string()),
  betterStructure: z.array(z.string()),
});

/** The whole completion screen in one payload — see docs/API-CONTRACT.md's
 * `GET /reports/{id}/completion`. */
export const sessionCompletionSchema = z.object({
  reportId: z.string(),
  sessionId: z.string(),
  code: z.string(),
  role: z.string(),
  company: z.string(),
  mode: z.string(),
  completedAt: z.string(),
  durationMinutes: z.number(),
  questionsAnswered: z.number(),
  overall: z.object({
    score: z.number(),
    band: z.string(),
    topPercent: z.number(),
    deltaFromPrevious: z.number(),
    caption: z.string(),
  }),
  summary: z.string(),
  signature: z.array(completionSignatureAxisSchema),
  metrics: z.array(completionMetricSchema),
  speech: speechMetricsSchema,
  strengths: z.array(z.string()),
  protocols: z.array(growthProtocolSchema),
  questions: z.array(completionQuestionSchema),
});

File: Backend/app/schemas/practice.py
from typing import TYPE_CHECKING, ClassVar

from app.core.serialization import CamelModel
from app.core.timeutils import UtcDatetime
from app.schemas.common import (
    AnswerVerdict,
    Difficulty,
    InterviewType,
    MetricTone,
    ProtocolPriority,
    SessionWireStatus,
)
from app.schemas.preparation import Question as QuestionRef

from app.schemas.interviewer import InterviewerLogEntry


class PracticeConfig(CamelModel):
    omit_if_none: ClassVar[frozenset[str]] = frozenset({"resume_id"})

    role: str
    company: str
    type: InterviewType
    difficulty: Difficulty
    duration: int
    focus_areas: list[str]
    interviewer_style: str
    resume_id: str | None = None


class SessionAnswer(CamelModel):
    omit_if_none: ClassVar[frozenset[str]] = frozenset({"follow_up"})

    question_id: str
    question: str
    transcript: str
    duration_seconds: int
    score: float
    strengths: list[str] = []
    missing: list[str] = []
    follow_up: bool | None = None


class PracticeSession(CamelModel):
    omit_if_none: ClassVar[frozenset[str]] = frozenset({"started_at", "interviewer_log"})

    id: str
    status: SessionWireStatus
    config: PracticeConfig
    questions: list[QuestionRef]
    current_question_index: int
    answers: list[SessionAnswer]
    started_at: UtcDatetime | None = None
    interviewer_log: list[InterviewerLogEntry] = []


class AnswerReview(CamelModel):
    question: str
    answer: str
    score: float
    strengths: list[str]
    missing: list[str]
    better_structure: list[str]


class SpeechMetrics(CamelModel):
    average_wpm: int
    filler_count: int
    fillers: dict[str, int]
    long_pauses: int
    longest_pause: float
    average_answer_seconds: int


class InterviewReport(CamelModel):
    id: str
    session_id: str
    created_at: UtcDatetime
    overall: int
    technical: int
    communication: int
    structure: int
    clarity: int
    relevance: int
    depth: int
    summary: str
    speech: SpeechMetrics
    weak_topics: list[str]
    strengths: list[str]
    recommended_actions: list[str]
    answers: list[AnswerReview]


class CompletionOverall(CamelModel):
    """The completion view's headline instrument — one score, what it means, and how it
    moved."""

    score: int
    band: str
    # Standing among comparable sessions, expressed as "top N%" — 3 reads as "TOP 3%".
    top_percent: int
    delta_from_previous: int
    caption: str


class SignatureAxis(CamelModel):
    """One spoke of the six-axis signature chart. `benchmark` is the target this axis is
    read against, not a peer average."""

    key: str
    label: str
    value: int
    benchmark: int


class CompletionMetric(CamelModel):
    """A metric tile. `band` is display copy for the value ("Optimal"), `tone` drives its
    colour only, and `delta` is already display-ready — null when there is no comparable
    previous session to measure against."""

    key: str
    label: str
    value: int
    band: str
    tone: MetricTone
    delta: str | None = None


class GrowthProtocol(CamelModel):
    id: str
    priority: ProtocolPriority
    title: str
    detail: str
    # Seeds the targeted-retry deep link back into /practice/setup.
    focus_area: str


class CompletionQuestion(CamelModel):
    """One asked question with its answer and per-answer analysis, as the completion
    view's question list renders it."""

    id: str
    position: int
    question: str
    topic: str
    category: str
    difficulty: Difficulty
    score: float
    duration_seconds: int
    verdict: AnswerVerdict
    answer: str
    strengths: list[str]
    missing: list[str]
    better_structure: list[str]


class SessionCompletion(CamelModel):
    """Everything the post-interview completion screen renders, composed from the report,
    the session that produced it, and the history log — see services/completion.py."""

    report_id: str
    session_id: str
    code: str
    role: str
    company: str
    mode: str
    completed_at: UtcDatetime
    duration_minutes: int
    questions_answered: int
    overall: CompletionOverall
    summary: str
    signature: list[SignatureAxis]
    metrics: list[CompletionMetric]
    speech: SpeechMetrics
    strengths: list[str]
    protocols: list[GrowthProtocol]
    questions: list[CompletionQuestion]


class AnswerCompletedRequest(CamelModel):
    omit_if_none: ClassVar[frozenset[str]] = frozenset({"pause_markers_ms"})

    question_id: str
    transcript: str
    started_at: UtcDatetime
    ended_at: UtcDatetime
    duration_ms: int
    pause_markers_ms: list[int] | None = None


class SocketTicket(CamelModel):
    ticket: str
    expires_at: UtcDatetime
